### PR TITLE
Report system stats with OSHI instead of libsigar

### PR DESCRIPTION
**Pull request by mpfz0r**
Improved version of this excellent PR from Mario Daniel Ruiz Saavedra (https://github.com/desiderantes)
https://github.com/Graylog2/graylog2-server/pull/6056
Thank you very much 🤝 

This commit adds a new Operating System and Hardware Information library, OSHI, that does not require the cumbersome packaging that Sigar currently needs. This library has a very active community and a pretty good set of platform options.

## Description
A new config option, `disable_native_system_stats_collector`, is added, defaults to false.
Implementations for FS, network, OS, and process probes.

## Motivation and Context
Sigar is pretty much unmaintained, and a replacement with a friendly license would be a nice addition.

## How Has This Been Tested?

Personally tested on Linux (Ubuntu 18.04) and MacOS 10.15.7

Compared the libsigar stats with oshi, which either match, or provide even more info

Disk statistics on MacOS need this fix https://github.com/oshi/oshi/pull/1464

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -147,7 +147,7 @@
         <scala.version>2.11.8</scala.version>
         <semver4j.version>2.2.0-graylog.1</semver4j.version>
         <shiro.version>1.5.2</shiro.version>
-        <sigar.version>1.6.4</sigar.version>
+        <oshi.version>5.3.6</oshi.version>
```
Comment (bernd) [inline pom.xml:150]: Can you please update to 5.3.7 release of oshi?

File: misc/graylog.conf [hunk 2]
```diff
@@ -637,8 +637,8 @@ mongodb_threads_allowed_to_block_multiplier = 5
 # Connection timeout for a configured LDAP server (e. g. ActiveDirectory) in milliseconds.
 #ldap_connection_timeout = 2000
 
-# Disable the use of SIGAR for collecting system stats
-#disable_sigar = false
+# Disable the use of OSHI for collecting system stats
+#disable_oshi = false
```
Comment (bernd) [inline misc/graylog.conf:641]: I think we should add this change to the UPGRADING.rst file.

Comment (bernd) [reply misc/graylog.conf:641]: And what do you think about using something like `disable_native_system_stats_collector` instead? That way we don't have to rename it again in the future. :wink:

File: graylog2-server/src/main/java/org/graylog2/shared/system/stats/fs/OshiFsProbe.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            oshiFileSystems.put(path,
+                    fileSystem.getFileStores().stream()
+                            .filter(fs -> path.startsWith(fs.getMount()))
+                            // We want the mountpoint closest to our location
+                            .max(Comparator.comparingInt(p -> Paths.get(p.getMount()).getNameCount()))
+                            .map(fs -> {
+                                // First try search for the diskstore with the logical volume or volume name
+                                Optional<HWDiskStore> diskStore = hardware.getDiskStores().stream()
+                                        .filter(ds -> ds.getName().equals(StringUtils.defaultIfEmpty(fs.getLogicalVolume(), fs.getVolume())))
+                                        .findFirst(); //.orElse(generateDummyDiskStore());
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/shared/system/stats/fs/OshiFsProbe.java:80]: Can we remove the commented code?

**Review by bernd — CHANGES_REQUESTED** [review]
A few comments.

I am also running into the following error when calling the `/system/stats` and `/system/stats/fs` endpoints. The other `/system/stats/*` endpoints work fine.

```
graylog_1        | 2021-01-13 17:33:23,142 ERROR: org.graylog2.shared.rest.exceptionmappers.AnyExceptionClassMapper - Unhandled exception in REST resource
graylog_1        | java.lang.ArithmeticException: / by zero                                          
graylog_1        |      at org.graylog2.shared.system.stats.fs.OshiFsProbe.lambda$fsStats$11(OshiFsProbe.java:114) ~[graylog.jar:?]
graylog_1        |      at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1321) ~[?:1.8.0_242]                    
graylog_1        |      at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169) ~[?:1.8.0_242]
graylog_1        |      at java.util.stream.ReferencePipeline$11$1.accept(ReferencePipeline.java:440) ~[?:1.8.0_242]                                                                         
graylog_1        |      at java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1699) ~[?:1.8.0_242]                                                                
graylog_1        |      at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482) ~[?:1.8.0_242]                                                                  
graylog_1        |      at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472) ~[?:1.8.0_242]                                                                              
graylog_1        |      at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708) ~[?:1.8.0_242]                                                              
graylog_1        |      at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:1.8.0_242]                           
graylog_1        |      at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566) ~[?:1.8.0_242]                         
graylog_1        |      at org.graylog2.shared.system.stats.fs.OshiFsProbe.fsStats(OshiFsProbe.java:106) ~[graylog.jar:?]               
graylog_1        |      at org.graylog2.shared.system.stats.StatsService.fsStats(StatsService.java:55) ~[graylog.jar:?]
graylog_1        |      at org.graylog2.rest.resources.system.StatsResource.fsStats(StatsResource.java:64) ~[graylog.jar:?]
graylog_1        |      at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_242]  
graylog_1        |      at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_242]
graylog_1        |      at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_242]
graylog_1        |      at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_242]              
graylog_1        |      at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[graylog.jar:?]                
graylog_1        |      at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255) [graylog.jar:?]
graylog_1        |      at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) [graylog.jar:?]                                                                                                   
graylog_1        |      at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) [graylog.jar:?]                                                                                                   
graylog_1        |      at org.glassfish.jersey.internal.Errors.process(Errors.java:292) [graylog.jar:?]                                                                                                  
graylog_1        |      at org.glassfish.jersey.internal.Errors.process(Errors.java:274) [graylog.jar:?]                                                                                                  
graylog_1        |      at org.glassfish.jersey.internal.Errors.process(Errors.java:244) [graylog.jar:?]                                                                                                  
graylog_1        |      at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) [graylog.jar:?]                                                                           
graylog_1        |      at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234) [graylog.jar:?]                                                                                      
graylog_1        |      at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) [graylog.jar:?]                                                                             
graylog_1        |      at org.glassfish.jersey.grizzly2.httpserver.GrizzlyHttpContainer.service(GrizzlyHttpContainer.java:356) [graylog.jar:?]                                                                                                                graylog_1        |      at org.glassfish.grizzly.http.server.HttpHandler$1.run(HttpHandler.java:200) [graylog.jar:?]
graylog_1        |      at com.codahale.metrics.InstrumentedExecutorService$InstrumentedRunnable.run(InstrumentedExecutorService.java:180) [graylog.jar:?]                                                                                                     graylog_1        |      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_242]
graylog_1        |      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_242]                                                                                                                                       graylog_1        |      at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
```

**Comment by bernd** [comment]
@juju2112 Once this PR is merged we can remove all `-Djava.library.path=../lib/sigar` (and similar) properties from startup scripts in 4.1 packages, docker containers, etc.

**Comment by juju2112** [comment]
> @juju2112 Once this PR is merged we can remove all `-Djava.library.path=../lib/sigar` (and similar) properties from startup scripts in 4.1 packages, docker containers, etc.

Awesome!

File: pom.xml (hunk 1 above)
Comment (mpfz0r) [reply pom.xml:150]: can do. but when I did this PR it wasn't even released :man_shrugging:

**Comment by mpfz0r** [comment]
> A few comments.
> 
> I am also running into the following error when calling the `/system/stats` and `/system/stats/fs` endpoints. The other `/system/stats/*` endpoints work fine.
> 

@bernd 
This is because we cannot find the right DiskStore..
Could you set a breakpoint here? https://github.com/Graylog2/graylog2-server/pull/9840/files#diff-b278a984e7ff2e46137b9e6d548dd66d33eea4714c7c0a0a39b004351eee763fR94

And then get me the output of `hardware.getDiskstores()`?

**Comment by mpfz0r** [comment]
> > A few comments.
> > I am also running into the following error when calling the `/system/stats` and `/system/stats/fs` endpoints. The other `/system/stats/*` endpoints work fine.
> 
> @bernd
> This is because we cannot find the right DiskStore..
> Could you set a breakpoint here? https://github.com/Graylog2/graylog2-server/pull/9840/files#diff-b278a984e7ff2e46137b9e6d548dd66d33eea4714c7c0a0a39b004351eee763fR94
> 
> And then get me the output of `hardware.getDiskstores()`?

nah, make that https://github.com/Graylog2/graylog2-server/pull/9840/files#diff-b278a984e7ff2e46137b9e6d548dd66d33eea4714c7c0a0a39b004351eee763fR67

and get me `fileSystem.getFileStores()`

**Comment by bernd** [comment]
> nah, make that https://github.com/Graylog2/graylog2-server/pull/9840/files#diff-b278a984e7ff2e46137b9e6d548dd66d33eea4714c7c0a0a39b004351eee763fR67
> 
> and get me `fileSystem.getFileStores()`

I sent that information outside of GitHub.

**Review by bernd — APPROVED** [review]
LGTM now! :+1:

**Pull request merged**